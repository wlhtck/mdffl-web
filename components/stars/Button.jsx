import React from 'react';
import {
  string, func, bool, oneOf,
} from 'prop-types';
import styled from 'styled-components';
import { primary, white } from '../nebulas/colors';

const Button = ({
  link, onClick, text, className, external,
}) => (
  link
    ? (
      <a
        className={className}
        href={link}
        target={external ? '_blank' : ''}
        rel={external ? 'noreferrer noopener' : ''}
      >
        {text}
      </a>
    )
    : <button className={className} onClick={onClick} type="button">{text}</button>

);

const backgroundColors = {
  primary,
  secondary: white,
};

const fontColors = {
  primary: white,
  secondary: primary,
};

Button.propTypes = {
  external: bool,
  link: string,
  onClick: func,
  text: string.isRequired,
  className: string,
  type: oneOf(['primary', 'secondary']),
};

Button.defaultProps = {
  external: false,
  link: '',
  onClick: null,
  className: '',
  type: 'primary',
};

export default styled(Button)`
  font-family: 'Raleway', sans-serif;
  font-size: 16px;
  background-color: ${({ type }) => backgroundColors[type]};
  color: ${({ type }) => fontColors[type]};
  display: inline-block;
  padding: 16px;
  border: none;
  line-height: 1;
  border-radius: 4px;
  cursor: pointer;
  text-decoration: none;
`;
