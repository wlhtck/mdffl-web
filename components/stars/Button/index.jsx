import {
  string, func, bool, oneOf,
} from 'prop-types';
import { omit } from 'lodash/fp';
import {
  compose, mapProps, componentFromProp, branch, withProps,
} from 'recompose';
import withStyles from '../../nebulas/withStyles';
import styles from './styles';

const omitProps = compose(mapProps, omit);
const withExternalLinkProps = withProps({ rel: 'noreferrer noopener', target: '_blank' });
const asButton = compose(omitProps(['href']), withProps(() => ({ component: 'button' })));
const asLink = compose(
  omitProps(['onClick']),
  withProps(() => ({ component: 'a' })),
  branch(
    ({ external }) => !!external,
    withExternalLinkProps,
  ),
);

const Button = compose(
  withStyles(styles),
  branch(
    ({ href }) => !href,
    asButton,
    asLink,
  ),
  omitProps(['external']),
)(componentFromProp('component'));

Button.propTypes = {
  children: string.isRequired,
  className: string,
  external: bool,
  onClick: func,
  type: oneOf(['primary', 'secondary']),
  url: string,
};

Button.defaultProps = {
  className: '',
  external: false,
  onClick: null,
  type: 'primary',
  url: null,
};

export default Button;
