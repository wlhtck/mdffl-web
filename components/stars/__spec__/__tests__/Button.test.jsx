import React from 'react';
import 'jest-styled-components';
import { render } from 'enzyme';
import Button from '../../Button';

describe('components/starts/Button', () => {
  it('renders an interal link', () => {
    expect(render(<Button href="/about">Button</Button>)).toMatchSnapshot();
  });

  it('renders an external link', () => {
    expect(render(<Button href="https://google.com" external>Button</Button>)).toMatchSnapshot();
  });

  it('renders a button', () => {
    expect(render(<Button onClick={() => {}}>Button</Button>)).toMatchSnapshot();
  });

  it('renders a secondary button', () => {
    expect(render(<Button type="secondary">Button</Button>)).toMatchSnapshot();
  });
});
