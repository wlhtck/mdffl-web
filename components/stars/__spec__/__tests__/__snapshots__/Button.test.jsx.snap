// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/starts/Button renders a button 1`] = `
<button
  class="withStyles-u5urts-0 gqQEcA"
  type="primary"
>
  Button
</button>
`;

exports[`components/starts/Button renders a secondary button 1`] = `
<button
  class="withStyles-u5urts-0 hpEXTr"
  type="secondary"
>
  Button
</button>
`;

exports[`components/starts/Button renders an external link 1`] = `
<a
  class="withStyles-u5urts-0 gqQEcA"
  href="https://google.com"
  rel="noreferrer noopener"
  target="_blank"
  type="primary"
>
  Button
</a>
`;

exports[`components/starts/Button renders an interal link 1`] = `
<a
  class="withStyles-u5urts-0 gqQEcA"
  href="/about"
  type="primary"
>
  Button
</a>
`;
